feat(insights): add custom function support to EAP

DominikB2014 opened the pull request:
Work for #81750 

1. Adds the ability to define custom functions (formulas)
2. Implement the `http_response_rate` function, which is used in insights

DominikB2014 committed:
implement response status code 5

Zylphrex commented on src/sentry/search/eap/columns.py:
```diff
@@ -117,6 +119,24 @@ class VirtualColumnDefinition:
 
 
+@dataclass(frozen=True, kw_only=True)
+class CustomFunction(ResolvedAttribute):
```
Maybe call this `ResolvedFormula`?

DominikB2014 replied:
👍

Zylphrex commented on src/sentry/search/eap/resolver.py:
```diff
@@ -36,10 +36,12 @@
     ResolvedFunction,
     VirtualColumnDefinition,
 )
+from sentry.search.eap.span_columns import CUSTOM_FUNCTIONS
```
You should pass this as part of the `ColumnDefinitions` as this file should remain independent of the dataset.

DominikB2014 replied:
fixed

Zylphrex commented on src/sentry/search/eap/resolver.py:
```diff
@@ -691,6 +694,9 @@ def resolve_aggregate(
         # Alias defaults to the name of the function
         alias = match.group("alias") or column
 
+        if function in CUSTOM_FUNCTIONS:
+            return (CUSTOM_FUNCTIONS.get(function), None)
```
I was thinking this could live within the existing `self.definitions.functions` as it's just a function that happens to resolve to a formula.

Then this could look more like

```python
function_definition = self.definitions.functions[function]
# handle the arguments here
resolved_aggregate = function_definition.resolve(...)  # this can return `ResolvedFunction | ResolvedFormula`
```

DominikB2014 replied:
fixed

DominikB2014 committed:
working function with args

DominikB2014 committed:
working function with args

DominikB2014 committed:
fix logic

DominikB2014 committed:
remove column

DominikB2014 committed:
updates

DominikB2014 committed:
typing

wmak commented on src/sentry/search/eap/span_formulas.py:
```diff
@@ -0,0 +1,109 @@
+    if code not in [1, 2, 3, 4, 5]:
+        raise Exception("http_response_rate takes in a single digit (1,2,3,4,5)")
+
+    response_code_map = {
```
Could we move these to a constants file

wmak commented on src/sentry/search/eap/span_formulas.py:
```diff
@@ -0,0 +1,109 @@
+
+    # TODO - handling valid parameters should be handled in the function_definitions (span_columns.py)
+    if code not in [1, 2, 3, 4, 5]:
+        raise Exception("http_response_rate takes in a single digit (1,2,3,4,5)")
```
We should stick with raising `InvalidSearchQuery` when its an invalid user query

wmak commented on src/sentry/search/eap/span_columns.py:
```diff
@@ -619,6 +620,19 @@ def count_processor(count_value: int | None) -> int:
+            )
+        ],
+        formula_resolver=CUSTOM_FUNCTION_RESOLVER["http_response_rate"],
+    ),
```
🤔 should we make a separate `SPAN_FORMULA_DEFINITIONS`? ie. if we're defining formulas in `span_formulas` maybe we just import the definitions from there and put it into the definitions directly?

wmak commented on src/sentry/search/eap/columns.py:
```diff
@@ -147,6 +145,54 @@ def proto_type(self) -> AttributeKey.Type.ValueType:
 
 
+@dataclass
+class FunctionDefinition:
```
Would it be possible to keep `FunctionDefinition` untouched and just introduce `FormulaDefinition`? since Functions will never have a `formula_resolver`?

DominikB2014 committed:
code review

DominikB2014 committed:
typing

DominikB2014 commented on src/sentry/search/eap/resolver.py:
```diff
@@ -692,9 +696,13 @@ def resolve_aggregate(
         # Get the function definition
-        if function not in self.definitions.functions:
+        function_definition: FunctionDefinition | FormulaDefinition
+        if function in self.definitions.functions:
+            function_definition = self.definitions.functions[function]
```
In a future refactor, we can rename `self.definitions.functions` to `self.definitions.aggregates`

DominikB2014 committed:
Merge branch 'master' into DominikB2014/eap-custom-functions

DominikB2014 commented:
@wmak I addressed your comments!
I slapped everything in one file. 
In future PRs, i'm planning to splitting up the `span_columns` as we discussed and address the TODO comments.

Lmk if that works for you!

wmak commented on src/sentry/search/eap/resolver.py:
```diff
@@ -324,7 +327,7 @@ def resolve_virtual_context_term(
         term: str,
         raw_value: str | list[str],
-        resolved_column: ResolvedColumn,
+        resolved_column: ResolvedColumn | ResolvedFormula | ResolvedFormula,
```
Is this intentional? VCCs should only apply to Resolved Columns

wmak commented on src/sentry/search/eap/resolver.py:
```diff
@@ -468,7 +471,7 @@ def resolve_aggregate_term(
     def _resolve_search_value(
         self,
-        column: ResolvedColumn,
+        column: ResolvedColumn | ResolvedFormula,
```
Another one, is this intentional? We don't resolve search values for formulas do we?

wmak commented on src/sentry/search/eap/span_columns.py:
```diff
@@ -386,6 +402,65 @@ def count_processor(count_value: int | None) -> int:
 
+def http_response_rate(arg: str) -> Column.BinaryFormula:
+
+    if not arg.isdigit():
```
nit: why not just check that arg is in a list of the strings of 1-5, then you can cast safely without concern?

wmak commented on src/sentry/search/eap/span_columns.py:
```diff
@@ -386,6 +402,65 @@ def count_processor(count_value: int | None) -> int:
+                        op=ComparisonFilter.OP_IN,
+                        value=AttributeValue(
+                            val_str_array=StrArray(
+                                values=response_codes,  #
```
```suggestion
                                values=response_codes,
```

wmak commented on src/sentry/search/eap/span_columns.py:
```diff
@@ -621,8 +696,24 @@ def count_processor(count_value: int | None) -> int:
+            ArgumentDefinition(
+                argument_types={
+                    "string"
+                },  # TODO - this should be an integer, but `resolve_attribute` returns a string
```
I think instead of leaving this as a todo we need to add a way to define arguments that aren't just attributes, so that `SearchResolver.resolve_aggregate` knows to cast a user input, which probably would clean up the definition of `http_response_rate` for you too

wmak reviewed (commented):
Overall in the right direction, I'm mostly concerned that we're sorta hacking the argument resolution currently.

DominikB2014 commented:
> Overall in the right direction, I'm mostly concerned that we're sorta hacking the argument resolution currently.

👍 thanks for the review, I don't mind spending some more time getting the args working properly

DominikB2014 committed:
code review

DominikB2014 committed:
code review

DominikB2014 commented on src/sentry/search/eap/columns.py:
```diff
@@ -147,6 +145,74 @@ def proto_type(self) -> AttributeKey.Type.ValueType:
+
+
+@dataclass
+class FormulaDefinition:
```
I think there's some class variables that don't apply to both. For example for `FormulaDefinition`, `internal_type` doesnt isn't relevant.

I'm thinking when we do the split, we could have a base class `FunctionDefinition`, and then `AggregateDefinition` and `FormulaDefinition` inherit from that?

DominikB2014 committed:
rename variable

DominikB2014 committed:
refactor

DominikB2014 commented:
@wmak I cleaned up the argument list a bit more. Since you looked there several changes
1. `ArgumentDefinition` optionally takes in `is_attribute` and `validator`
2. A couple variable name changes so things make more sense (previously everything assumed an arg was an attribute)
3. `resolve_aggregate` iterates over the arguments we now
a. checks against the argument validator, if the argument validator fails the query fails
b. If the argument is not an attribute, we skip `resolve_attribute` and just parse the attribute directly based on the `argument_types` (We don't have a use case for multiple argument types in formulas yet, so i didn't handle it in this PR)
4. Now that the FormulaDefinition defines all the validation, there is no need to validate in the `formula_resolver` itself
5. Now `resolve_aggregate` parses the argument, the `formula_resolver` no longer needs to do any parses and assumes the params are as expected

DominikB2014 committed:
fix

DominikB2014 committed:
Merge remote-tracking branch 'origin/master' into DominikB2014/eap-custom-functions

DominikB2014 committed:
comment

wmak commented on src/sentry/search/eap/resolver.py:
```diff
@@ -692,24 +696,47 @@ def resolve_aggregate(
-            if index < len(attribute_args):
-                parsed_argument, _ = self.resolve_attribute(attribute_args[index])
+            if argument.validator is not None:
+                if not argument.validator(arguments[index]):
+                    raise InvalidSearchQuery(
+                        f"{arguments[index]} is not a valid argument for {function}"
+                    )
```
nit(can be a follow up, or a TODO comment in code): can we change this so the `validator` is the one in charge of raising its error? That way we can give more specific errors?

DominikB2014 replied:
The validator does raise the error here, I just put this in to make it more robust. If someone mistakenly forgets to raise something, it will still get raised here. Maybe there's a way to enforce the validator raising the error?

https://github.com/getsentry/sentry/pull/85662/files#diff-8c5ca2161049ebb16b08be9488c6eec28f895ce8dce39b970a1af78bf3ead312R452

wmak reviewed (approved):
As discussed in slack i still want to clean up the FormulaDefinition & FunctionDefinition, but those are fine in a follow up. Otherwise this lgtm
